realtek: add rt-loader

plappermaul opened the pull request:
The bootloader of many Realtek switches only supports gzipped kernel images. With limited flash space that might get critical in future versions. For better compression allow support for compressed images. For this a new loader was developed. Several ideas have been taken over from the existing lzma loader but this has been enhanced to make integration simpler. What is new:

- Loader is position independent. No need to define load addresses
- Loader identifies device memory on its own
- Loader uses "official" upstream kernel lzma uncompress https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/decompress_unlzma.c
- Loader uses "official" UNMODIFIED nanoprintg that is used by several bare metal projects. https://github.com/charlesnicholson/nanoprintf

Compiled the loader ist just under 12KiB and during boot it will show:

```
rt-loader
Found RTL8380M (chip id 6275C) with 256MB
Relocate 2924240 bytes from 0x80100000 to 0x8fce0000
Extract kernel with 2900144 bytes from 0x8fce521c to 0x80100000...
Extracted kernel size is 9814907 bytes
Booting kernel from 0x80100000 ...

[    0.000000] Linux version 6.12.33 ...
[    0.000000] RTL838X model is 83806800
...
```

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
@achterin and while you are testing https://github.com/openwrt/openwrt/pull/18402 feel free to add this PR to your tests.

A successfull boot from initramfs **and from flash** is enough.

plappermaul commented:
I forgot:

- a RTL83xx device is rquired
- you need to add the LZMA recipe to your device in the Makefile (see commit for LGS310C)

achterin commented:
> @achterin and while you are testing #18402 feel free to add this PR to your tests.
> 
> A successfull boot from initramfs **and from flash** is enough.

sadly i can only offer to test on a zyxel xgs-1250-12 which is rtl930x

Borromini commented:
I will test when I find the time, but pressing real life issues at hand now.

Borromini commented:
@plappermaul Can you modify the PNR to include rtl930x already? I can test on XGS1250-12 v1 and XGS1210-12 v2.

plappermaul commented:
Off you go ...

Borromini commented:
Thanks. I can confirm it works on an XGS1250-12 v1, both ramdisk and flashed image.

@achterin You can test as follows on the XGS1250-12:

```
--- a/target/linux/realtek/image/rtl930x.mk
+++ b/target/linux/realtek/image/rtl930x.mk
@@ -21,6 +21,7 @@ endef
 TARGET_DEVICES += xikestor_sks8300-8x
 
 define Device/zyxel_xgs1250-12
+  $(Device/uimage-lzma-loader)
   SOC := rtl9302
   UIMAGE_MAGIC := 0x93001250
   ZYXEL_VERS := ABWE
@@ -31,8 +32,8 @@ define Device/zyxel_xgs1250-12
   KERNEL_INITRAMFS := \
        kernel-bin | \
        append-dtb | \
-       gzip | \
+       lzma | \
        zyxel-vers | \
-       uImage gzip
+       uImage lzma
 endef
 TARGET_DEVICES += zyxel_xgs1250-12

```

Borromini commented:
@plappermaul Any advice on how to migrate the GS1900 series? It relies on a common `zyxel_gs1900` recipe in target/linux/realtek/image/common.mk`. The libdeflate part doesn't seem to have a readymade lzma counterpart when grepping the tree.

```
define Device/zyxel_gs1900
  DEVICE_COMPAT_VERSION := 2.0
  DEVICE_COMPAT_MESSAGE := Dual firmware paritition merged due to size constraints. \
        Upgrade requires a new factory install. Regular sysupgrade is not possible.
  DEVICE_VENDOR := Zyxel
  IMAGE_SIZE := 13952k
  UIMAGE_MAGIC := 0x83800000
  KERNEL_INITRAMFS := \
        kernel-bin | \
        append-dtb | \
        libdeflate-gzip | \
        zyxel-vers | \
        uImage gzip | \
        check-size 6976k
endef
```

plappermaul commented:
lzma is a drop-in for libdeflate. Without patch all resolves to

```
KERNEL := kernel-bin | append-dtb | libdeflate-gzip | ...
KERNEL_INITRAMFS := kernel-bin | append-dtb | libdeflate-gzip | ...
```

With patch that gets

```
KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | ...
KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | loader-kernel | ...
```

For the bootloader to recognize "uimage gzip" mus be replaced by "uimage none"

Btw: Can you officially approve this patch in the commits section?

Borromini commented:
Wouldn't you like testing on rtl83xx first?

janh commented:
Works on XGS1010-12 v1 (tested only with initramfs image loaded via tftpboot so far). However, on this particular device, a regular LZMA-compressed image works as well.

Maybe I can do some further tests later (flash and rtl838x/rtl839x). But my rtl838x/rtl839x devices are all HPE 1920, which already use LZMA (via 7z). So it probably wouldn't make sense to actually use lzma-loader on those.

plappermaul commented:
@janh  Each tests helps. So thanks for that. Please approve

@Borromini I have no experience with LZMA loader. So starting with an a per-device approach.

Borromini commented:
@janh Would you mind sharing how you alter the build recipe for the XGS1010-12 to boot a LZMA compressed kernel? Because as you can see in the topic plappermaul linked to earlier, I had no luck with that.

@plappermaul My diff above for the GS1900 common recipe does not work, kernel hangs.

plappermaul commented:
How far do you get? Do you at least see the lzma loader messages?

janh commented:
@Borromini

> Would you mind sharing how you alter the build recipe for the XGS1010-12 to boot a LZMA compressed kernel?

This is what works on my XGS1010-12 v1:
https://github.com/janh/openwrt/blob/e6254cd2915a95a294cce54ab3f4953abe78c47b/target/linux/realtek/image/rtl930x.mk#L23-L41

It looks like your recipe in the linked post is correct, but the bootloader doesn't like the image.

> My diff above for the GS1900 common recipe does not work, kernel hangs.

With these changes, the initramfs image is LZMA compressed, while the header says it is uncompressed. But it lacks the LZMA loader.

I think something like this should work for GS1900:
```
  KERNEL_INITRAMFS := \
	$$(KERNEL/lzma-loader) | \
	zyxel-vers | \
	uImage none | \
	check-size 6976k
```

janh commented on target/linux/realtek/image/Makefile:
```diff
@@ -3,9 +3,14 @@
+
+LOADER_LOADADDR := 0x80100000
+LOADER_PLATFORM := realtek
+LOADER_FLASH_START := 0xb4000000
```
This seems unnecessary, because we don't build the loader in OKLI mode.

plappermaul replied:
Thanks for the review. Removed that part, simplified the integration and pushed a new version.

plappermaul commented:
@Borromini loader based lzma tftpboot works fine for me on XGS1210. Recipe:

```
define Device/zyxel_xgs1210-12
  $(Device/uimage-lzma-loader)
  SOC := rtl9302
  UIMAGE_MAGIC := 0x93001210
  ZYXEL_VERS := ABTY
  DEVICE_VENDOR := Zyxel
  DEVICE_MODEL := XGS1210-12
  IMAGE_SIZE := 13312k
  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | loader-kernel | zyxel-vers $$$$(ZYXEL_VERS) | \
       uImage none
endef
TARGET_DEVICES += zyxel_xgs1210-12
```

log

```
RTL9300# # rtk network on
Enable network
RTCORE Driver Module Initialize
  IOAL init
  Hardware-profile probe  GPIO probe (unit 0): (found)
  GPIO Init
  RTL8231 probe (unit 0): (found)
  RTL8231 init (unit 0)
 (XGS1210_12)
  Hardware-profile init
  GPIO probe (unit 0): (found)
  GPIO Init rtl9300_gpio_init had already been initialized!

  SPI init (unit 0)
  I2C probe (unit 0)
  I2C init (unit 0)
  RTL8231 probe (unit 0): (found)
  RTL8231 init (unit 0)
 r9300_rtl8231_init had already been initialized!
  NIC probe (unit 0)
  Loader RTNIC Driver Module Initialize
  IOAL init
RTK Driver Module Initialize
  MAC probe (unit 0)
    Chip 9302 (found)
  MAC init (unit 0)
  SMI protocol probe (unit 0)
  PHY probe (unit 0)
  Chip Construct (unit 0)
    Chip Construct
    Disable PHY Polling
    PHY Reset
    MAC Construct
    Turn Off Serdes
    Serdes Construct
    PHY Construct
    Turn On Serdes
    Enable PHY Polling
    Misc
  PHY init (unit 0)
  Mgmt_dev init (unit 0)
Please wait for PHY init-time ...

RTL9300# # tftpboot 0x8f000000 192.168.2.86:openwrt-realtek-rtl930x-zyxel_xgs1210-12-initramfs-kernel.bin
Using rtl9300#0 device
TFTP from server 192.168.2.86; our IP address is 192.168.2.71
Filename 'openwrt-realtek-rtl930x-zyxel_xgs1210-12-initramfs-kernel.bin'.
Load address: 0x8f000000
Loading: #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ########################
done
Bytes transferred = 5119184 (4e1cd0 hex)
RTL9300# # bootm
## Booting kernel from Legacy Image at 8f000000 ...
   Image Name:   MIPS OpenWrt Linux-6.6.85
   Created:      2025-04-05  12:20:15 UTC
   Image Type:   MIPS Linux Kernel Image (uncompressed)
   Data Size:    5119120 Bytes = 4.9 MB
   Load Address: 80100000
   Entry Point:  80100000
   Verifying Checksum ... OK
   Loading Kernel Image ... OK
OK

Starting kernel ...



OpenWrt kernel loader for MIPS based SoC
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Decompressing kernel... done!
Starting kernel at 80100000...

[    0.141924] printk: console [ttyS0] enabled
[    0.146676] printk: bootconsole [early0] disabled
[    0.209585] brd: module loaded
[    0.227118] spi-nor spi0.0: mx25l12805d (16384 Kbytes)
[    0.232960] 7 fixed-partitions partitions found on MTD device spi0.0
[    0.240129] OF: Bad cell count for /soc/spi@1200/flash@0/partitions
[    0.247167] OF: Bad cell count for /soc/spi@1200/flash@0/partitions
[    0.254385] Creating 7 MTD partitions on "spi0.0":
[    0.259785] 0x000000000000-0x0000000e0000 : "u-boot"
[    0.306693] 0x0000000e0000-0x0000000f0000 : "u-boot-env"
[    0.326567] 0x0000000f0000-0x000000100000 : "u-boot-env2"
[    0.333981] 0x000000100000-0x000000200000 : "jffs"
[    0.346609] 0x000000200000-0x000000300000 : "jffs2"
[    0.366594] 0x000000300000-0x000000fe0000 : "firmware"
[    0.375801] 0x000000fe0000-0x000001000000 : "log"
[    0.407560] Probing RTL838X eth device pdev: 818a7200, dev: 818a7210
[    0.433862] Found SoC ID: 9302: RTL9302B, family 9300
[    0.439663] Using MAC 000000e04c000000
```

achterin commented:
> Thanks. I can confirm it works on an XGS1250-12 v1, both ramdisk and flashed image.
> 
> @achterin You can test as follows on the XGS1250-12:
> 
> ```
> --- a/target/linux/realtek/image/rtl930x.mk
> +++ b/target/linux/realtek/image/rtl930x.mk
> @@ -21,6 +21,7 @@ endef
>  TARGET_DEVICES += xikestor_sks8300-8x
>  
>  define Device/zyxel_xgs1250-12
> +  $(Device/uimage-lzma-loader)
>    SOC := rtl9302
>    UIMAGE_MAGIC := 0x93001250
>    ZYXEL_VERS := ABWE
> @@ -31,8 +32,8 @@ define Device/zyxel_xgs1250-12
>    KERNEL_INITRAMFS := \
>         kernel-bin | \
>         append-dtb | \
> -       gzip | \
> +       lzma | \
>         zyxel-vers | \
> -       uImage gzip
> +       uImage lzma
>  endef
>  TARGET_DEVICES += zyxel_xgs1250-12
> ```

Successfully tested on Zyxel XGS-1250-12.

Borromini commented:
> How far do you get? Do you at least see the lzma loader messages?

Nope:

```
TFTP from server 10.0.0.10; our IP address is 10.0.0.100
Filename 'openwrt-realtek-rtl838x-zyxel_gs1900-8hp-v2-initramfs-kernel.bin'.
Load address: 0x84f00000
Loading: #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #########################
done
Bytes transferred = 5127295 (4e3c7f hex)
RTL838x# bootm
   Version:   MIPS OpenWrt Linux-6.6.85
   Created:   2025-04-06   7:48:39 UTC
   Size:      5127231 Bytes = 4.9 MB
   Verifying Checksum ... OK
   Loading Image ... OK
OK

Starting ...
```

plappermaul commented:
So you use todays PR version and this recipe in common.mk ?

```
define Device/zyxel_gs1900
  ...
  KERNEL_INITRAMFS := \
        kernel-bin | \
        append-dtb | \
        lzma | \
        loader-kernel | \
        zyxel-vers | \
        uImage none | \
        check-size 6976k

```

janh reviewed (approved):
Tested with initramfs image on HPE 1920-8G (rtl838x), HPE 1920-48G (rtl839x), Zyxel XGS1010-12 (rtl930x). Works fine on all of them (but none of them actually need it).

<details>
<summary>Used recipe for XGS1010-12</summary>

```Makefile
define Device/zyxel_xgs1010-12
  $(Device/uimage-lzma-loader)
  SOC := rtl9302
  UIMAGE_MAGIC := 0x93001010
  DEVICE_VENDOR := Zyxel
  DEVICE_MODEL := XGS1010-12
  KERNEL_SIZE := 7168k
  IMAGE_SIZE := 13184k
endef
TARGET_DEVICES += zyxel_xgs1010-12
```

</details>

<details>
<summary>Used patch for HPE 1920</summary>

```diff
diff --git a/target/linux/realtek/image/Makefile b/target/linux/realtek/image/Makefile
index 19fab03dba..ca13eba1a0 100644
--- a/target/linux/realtek/image/Makefile
+++ b/target/linux/realtek/image/Makefile
@@ -60,7 +60,7 @@ define Build/h3c-image
        $(STAGING_DIR_HOST)/bin/mkh3cimg \
                -i $(@) \
                -o $(@).new \
-               -c 7z \
+               -c none \
                -p $(H3C_PRODUCT_ID) \
                -d $(H3C_DEVICE_ID)
        mv $@.new $@
diff --git a/target/linux/realtek/image/common.mk b/target/linux/realtek/image/common.mk
index e600b2347d..0e596821b3 100644
--- a/target/linux/realtek/image/common.mk
+++ b/target/linux/realtek/image/common.mk
@@ -32,6 +32,7 @@ define Device/d-link_dgs-1210
 endef
 
 define Device/hpe_1920
+  $(Device/uimage-lzma-loader)
   DEVICE_VENDOR := HPE
   IMAGE_SIZE := 29632k
   BLOCKSIZE := 64k
@@ -39,13 +40,15 @@ define Device/hpe_1920
   KERNEL := \
        kernel-bin | \
        append-dtb | \
-       7z | \
+       lzma | \
+       loader-kernel | \
        h3c-image | \
        h3c-vfs
   KERNEL_INITRAMFS := \
        kernel-bin | \
        append-dtb | \
-       7z | \
+       lzma | \
+       loader-kernel | \
        h3c-image
   IMAGE/sysupgrade.bin := \
        append-kernel | \
```

</details>

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -3,7 +3,7 @@
 include $(INCLUDE_DIR)/image.mk
 
-KERNEL_LOADADDR = 0x80100000
+KERNEL_LOADADDR := 0x80100000
```
Do you need to change the assignment type here for any reason? Without variables in the assignement, the `:=` assignment would only serve to make things marginally more efficient AFAICT (not re-evaluating the expression for every use).

plappermaul replied:
Simply following the standard

```
[markus@openwrt linux]$ grep KERNEL_LOADADDR */image/Makefile | grep " := " | wc -l
18
[markus@openwrt linux]$ grep KERNEL_LOADADDR */image/Makefile | grep " = " | wc -l
4

```

plappermaul replied:
Fine with that?

svanheule commented on target/linux/realtek/image/Makefile:
```diff
@@ -85,6 +103,7 @@ define Build/linksys-image
 endef
 
 define Device/Default
+  LZMA_TEXT_START := 0x82000000
```
What are the typical sizes of the compressed (loader + payload) and decompressed (payload) blobs? If an uncompressed payload is already close to 32MB, might it make sense to move this to e.g. 48MB?

plappermaul replied:
Simply going with the "standard".

```
[markus@openwrt linux]$ grep KERNEL_LOADADDR */image/Makefile | grep " := " | wc -l
18
[markus@openwrt linux]$ grep KERNEL_LOADADDR */image/Makefile | grep " = " | wc -l
4
```

plappermaul replied:
No real idea. just following the largest value of other targets here:

```
[openwrt linux]$ grep LZMA_TEXT_START  */image/Makefile | grep 0x
ath79/image/Makefile:  LZMA_TEXT_START := 0x81800000
bmips/image/Makefile:LZMA_TEXT_START := 0x82000000              # RAM start + 32M
lantiq/image/Makefile:          LZMA_TEXT_START=0x82000000 \
ramips/image/Makefile:  LZMA_TEXT_START := 0x81800000
realtek/image/Makefile:  LZMA_TEXT_START := 0x82000000

```

plappermaul replied:
Fine with that?

svanheule reviewed (commented):
Thanks for the patches. Better compression support would be easier for users wanting to keep dual-boot on devices with small flash.

Do you happen to have an idea of how this compares to a self-extracting kernel?

Perhaps more for other devices that still have enough room for bigger images: How much impact does this have on boot times?

Neustradamus commented:
To follow this PR :)

undisputed-seraphim commented:
Excited for this. I am slowly chipping away at RTL8198C experiments and upstreaming this patch makes things more convenient!

plappermaul committed:
realtek: add rt-loader (runtime loader)

The bootloader of many Realtek switches only supports gzipped kernel images.
With limited flash space that might get critical in future versions. For better
compression allow support for compressed images. For this a new loader was
developed. Several ideas have been taken over from the existing lzma loader
but this has been enhanced to make integration simpler. What is new:

- Loader is position independent. No need to define load addresses
- Loader identifies device memory on its own
- Loader uses "official" upstream kernel lzma uncompress
  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/decompress_unlzma.c
- Loader uses "official" UNMODIFIED nanoprintg that is used by several
  bare metal projects. https://github.com/charlesnicholson/nanoprintf

Compiled the loader ist just under 12KiB and during boot it will show:

rt-loader
Found RTL8380M (chip id 6275C) with 256MB
Relocate 2924240 bytes from 0x80100000 to 0x8fce0000
Extract kernel with 2900144 bytes from 0x8fce521c to 0x80100000...
Extracted kernel size is 9814907 bytes
Booting kernel from 0x80100000 ...

[    0.000000] Linux version 6.12.33 ...
[    0.000000] RTL838X model is 83806800
...

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18397
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: add rt-loader recipe

To make use of the new rt-loader provide the needed recipes.
This has been tested with the following devices:

- rtl838x Linksys LGS310: initramfs & flash
- rtl930x Zyxel XGS1210: initramfs

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18397
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul committed:
realtek: Convert LGS310C to compressed kernel

There are too many supported Realtek devices so avoid activating the
rt loader recipe in the default builds. Just start with the LGS310C.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18397
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented:
Always disliked the existing lzma loader. Created a new xz loader from scratch to make life much easier ...

BrainSlayer commented:
> I forgot:
> 
> * a RTL83xx device is rquired
> * you need to add the LZMA recipe to your device in the Makefile (see commit for LGS310C)

why is rtl83xx required? because if your specific ram detection which isnt neccessary? the lzma loader you used before and the one i'm using works on all chipsets

Djfe commented on target/linux/realtek/image/xz-loader/src/memory.c:
```diff
@@ -0,0 +1,113 @@
+ * xz-loader memory functions
+ * (c) 2025 Markus Stockhausen
+ *
+ * This is a small function collection to get some rudimentary memory managemenet working when
```
just a typo, I noticed: management

plappermaul replied:
Fixed

plappermaul commented on target/linux/realtek/image/xz-loader/src/startup.S:
```diff
@@ -0,0 +1,182 @@
+#
+# This start code allows to run a position independent code (PIC) on bare metal. In that case
+# all addresses are looked up via the global offset table (GOT). But that must be filled during
+# this initialization sequence. Without a porper GOT using standard "la" instruction in the code
```
Fixed

plappermaul commented:
You are right. xz is only an additional header with extra checksum. But if you look at the current toolchain we see another indication why lzma should be avoided. 

./host/bin/lzma
LZMA 4.65 : Igor Pavlov : Public domain : 2009-02-03
Usage:  LZMA <e|d> inputFile outputFile [<switches>...]

This is no excuse. Just want to differentate from lzma-loader to make clear that this is something different

plappermaul commented:
Ok to sum up the current discussion at this point.

- We are discussing about the compressed format "xz vs lzma".
- But the real benefit is having an auto tuning loader and a better compression than gzip
- So to save time for fixing other bugs I would like to go with the current implmentation

What did I do now?

- I just rebased the xz loader on top of current main
- Changed the printf library from mpaland/printf to eyalroz/printf as this is the successor project.
- Verified that everything is still working on 6.12

So I just ask everyone who can spend some time to run a short retest (inkl. boot from flash) with their device. 

Thanks.

BrainSlayer commented:
> Ok to sum up the current discussion at this point.
> 
> * We are discussing about the compressed format "xz vs lzma".
> * But the real benefit is having an auto tuning loader and a better compression than gzip
> * So to save time for fixing other bugs I would like to go with the current implmentation
> 
> What did I do now?
> 
> * I just rebased the xz loader on top of current main
> * Changed the printf library from mpaland/printf to eyalroz/printf as this is the successor project.
> * Verified that everything is still working on 6.12
> 
> So I just ask everyone who can spend some time to run a short retest (inkl. boot from flash) with their device.
> 
> Thanks.

xz vs lzma -> lzma format is smaller
xz vs lzma code -> lzma decoder is smaller since its more simple designed
and finally. i really dont care. its just a little bit waste of resources to reinvent the wheel if the result has negative effects.
keep it simple. keep it small. or as we said in the 90s. a well written borland c code always starts with "asm {"
just joking in fact its mov al,13/ int 10

Djfe commented:
Not that it matters to me (I don't want to give fuel into the fire).
But here was some discussion/insight into current view points of maintainers on xz / lzma vs lzip etc.
This was right after the incident, where xz was abused as a backdoor on hosts running systemd last year.

I just thought it might be an interesting read, even though it was more on the topic of how to compress packages. They also talk about how boot images will remain xz for a while since bootloaders expect the format.
https://github.com/openwrt/openwrt/commit/bab3ae2ee7656600a185f4cef11cef94389023af#commitcomment-140694079

also the link by robimarko two comments above the one I just linked: https://www.nongnu.org/lzip/xz_inadequate.html

I don't have any preference either way on what should be used. But maybe this gives either side some arguments.
I most definitely don't want to invalidate anyone's work or question it.

BrainSlayer commented:
> Not that it matters to me (I don't want to give fuel into the fire). But here was some discussion/insight into current view points of maintainers on xz / lzma vs lzip etc. This was right after the incident, where xz was abused as a backdoor on hosts running systemd last year.
> 
> I just thought it might be an interesting read, even though it was more on the topic of how to compress packages. They also talk about how boot images will remain xz for a while since bootloaders expect the format. [bab3ae2#commitcomment-140694079](https://github.com/openwrt/openwrt/commit/bab3ae2ee7656600a185f4cef11cef94389023af#commitcomment-140694079)
> 
> also the link by robimarko two comments above the one I just linked: https://www.nongnu.org/lzip/xz_inadequate.html
> 
> I don't have any preference either way on what should be used. But maybe this gives either side some arguments. I most definitely don't want to invalidate anyone's work or question it.
> 
> EDIT: I agree with the sentiment that this should be merged/appears to be an improvement. Further changes can be made/discussed in a new PR which changes the compression format/gives further options. Why? Mostly because it can take months for such a PR to be approved/merged and this improvement could be used until then.

not a single bootloader does expect or support xz. they use gzip or lzma and xz uses lzma or lzma2 lzma2 is basicly lzma + some binary filters to presort data for better compression. and xz puts then just a additional header in front. the backdoor in xz bzw was not inside the algorithm or code. it was part of the makefile/configure system

plappermaul commented:
A bootlader expects what it is made for. Be it xz or lzma or something else. But it is always better tohave one than to need one. Attached you will find my latest version. I hope you can make friends with it so that I can take care of more important things. If yes feel free to approve.

plappermaul commented:
@robimarko currently testing a a lot of PHY initialization without UBoot pre-init via serial uploads. Getting this merged would be a relief

Djfe commented:
not that I disagree with the notion but why aren't you already building changes on top of this branch/3 commits then? :)

robimarko commented:
Thanks! Rebased on top of main and merged!
